src/main/java/ch/njol/skript/lang/function/Function.java

```diff
@@ -73,7 +79,91 @@ public boolean isSingle() {
 	public ClassInfo<T> getReturnType() {
 		return sign.getReturnType();
 	}
-	
+
+	/**
+	 * Executes the function with the given event and parameters.
+	 * @param event The event for the function execution.
+	 * @param rawParameters The parameters for the function execution.
+	 * @return The result(s) of the function execution.
+	 * Returns null if the parameters are invalid or cannot be processed.
+	 */
+	public final T @Nullable [] execute(Event event, @Nullable Expression<?> rawParameters) {
+		Expression<?>[] parameters;
+		if (rawParameters instanceof ExpressionList<?>) {
+			ExpressionList<?> list = (ExpressionList<?>) rawParameters;
+			Expression<?>[] expressions = list.getExpressions();
+			parameters = list.getAnd() ? expressions : new Expression[] {CollectionUtils.getRandom(expressions)};
+		} else {
+			parameters = rawParameters != null ? new Expression[] {rawParameters} : new Expression[0];
+		}
+		return execute(event, parameters);
+	}
+
+	/**
+	 * Executes the function with the given event and parameters.
+	 * @param event The event for the function execution.
+	 * @param parameters The parameters for the function execution.
+	 * @return The result(s) of the function execution.
+	 * Returns null if the parameters are invalid or cannot be processed.
+	 */
+	public final T @Nullable [] execute(Event event, Expression<?>[] parameters) {
+		if (!validateParameters(parameters))
+			return null;
+		Object[][] params = processParameters(event, parameters);
+		if (params == null)
+			return null;
+		return execute(params);
+	}
+
+	@SuppressWarnings("unchecked")
+	private boolean validateParameters(Expression<?>[] parameters) {
+		if (parameters.length < sign.getMinParameters())
+			return false;
+
+		boolean singleListParam = sign.getMaxParameters() == 1 && !getParameter(0).single;
+		if (!singleListParam && parameters.length > sign.getMaxParameters())
+			return false;
+
+		for (int i = 0; i < parameters.length; i++) {
+			Parameter<?> parameter = sign.parameters[singleListParam ? 0 : i];
+			Expression<?> expression = parameters[i].getConvertedExpression(parameter.type.getC());
+			if (expression == null || (parameter.single && !expression.isSingle()))
+				return false;
+			parameters[i] = expression;
+		}
+		return true;
+	}
+
+	private Object @Nullable [][] processParameters(Event event, Expression<?>[] parameters) {
+		boolean singleListParam = sign.getMaxParameters() == 1 && !getParameter(0).single;
+
+		Object[][] params = new Object[singleListParam ? 1 : parameters.length][];
+		if (singleListParam && parameters.length > 1) { // All parameters to one list
+			List<Object> l = new ArrayList<>();
```

Comment:
Let's use better naming here please.

Reply:
I just moved the validation code from FunctionReference and moved it here. I guess I could clean it up as well

Reply:
That would be nice :)